### PR TITLE
[Feature] Support GLM-5.2 source-layer DSA indexer training

### PR DESCRIPTION
**Pull request by sjyouuuuug**
## Motivation

  Enable training of GLM-5.2 source-layer DSA indexers with attention-derived KL supervision.

  ## Changes

  - Add a cuDNN DSA indexer KL-loss operator and custom backward integration.
  - Separate differentiable indexer feature projection from no-grad Top-k selection.
  - Construct sparse attention teacher and indexer student distributions.
  - Aggregate source-layer indexer losses at the model boundary.
  - Keep physical MTP indexers frozen.
  - Add a GLM-5.2 training configuration and launch script.
  - Add tests for KL construction, gradient boundaries, score scaling, query masking, source/shared-layer behavior, and MTP freezing.

  ## Validation

  - Indexer loss CPU tests: 10 passed.
  - Source-indexer attention tests: 10 passed.
  - Multi-GPU tiny training completed all configured training steps.
  - Indexer loss decreased during the 300-step training run.

  ## Current limitations

  - IndexShare reuses source-layer Top-k indices, but shared-layer teacher supervision is not included.
  - Indexer training currently requires:
    - `SP_SIZE=1`
    - `INTRA_LAYER_MICRO_BATCH=1`
    - the `cudnn_dsa` backend
 
  ## Results
5cad86b0b8795b336b4f04cfd95114c
image

### CODE REVIEW

File: xtuner/v1/module/attention/dsa_mla.py [hunk 1]
```diff
@@ -141,10 +173,14 @@ def forward(
         # q: [bsz, S, Ni, Di]; k: [bsz, S, Di]
         q = torch.cat([q_pe, q_nope], dim=-1)
         k = torch.cat([k_pe, k_nope], dim=-1)
-        # weights: [bsz, S, Ni]
-        weights = self.weights_proj(hidden_states).float() * (self.index_n_heads**-0.5)
-
-        # Top-k 索引是整数，不需要梯度，所以整个 indexer 都放在 no_grad 下。
+        raw_weights = self.weights_proj(hidden_states)
```
Comment (jayhenry) [inline xtuner/v1/module/attention/dsa_mla.py]: 两个weights能否统一？

File: xtuner/v1/module/attention/dsa_mla.py [hunk 2]
```diff
@@ -365,17 +430,50 @@ def forward(
...
+            # The indexer learns from attention, but must not inject an extra
+            # gradient path into the transformer hidden/Q-LoRA activations.
+            indexer_features = self.indexer.project_features(
+                hidden_states.detach(),
+                q_resid.detach(),
                 position_embeddings,
                 seq_ctx,
-            ),
-        )
+            )
```
Comment (jayhenry) [inline xtuner/v1/module/attention/dsa_mla.py:564]: 对于 share layer 则不训练 indexer？

File: xtuner/v1/ops/sparse_mla/cudnn_dsa_indexer_loss.py [hunk 3]
```diff
@@ -0,0 +1,614 @@
...
+    topk_indices: Tensor,
+    *,
+    softmax_scale: float,
+    row_coefficient: float,
+    topk_length: Tensor | None = None,
+    valid_query_mask: Tensor | None = None,
+    debug_name: str | None = None,
+    debug_interval: int = 0,
+) -> Tensor:
+    """Convenience wrapper for one attention teacher and one indexer."""
```
Comment (jayhenry) [inline xtuner/v1/ops/sparse_mla/cudnn_dsa_indexer_loss.py:544]: 是否有参考来源？

Comment (sjyouuuuug) [reply xtuner/v1/ops/sparse_mla/cudnn_dsa_indexer_loss.py:544]: 主要参考 https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/transformer/experimental_attention_variant/dsa.py 中的compute_dsa_indexer_loss，算法的逻辑上是比较近似的，主 attention 的 query 计算 teacher attention distribution，通过 KL(teacher || indexer) 训练 indexer。sparse 模式下使用 Top-k mask 限制 token。

File: xtuner/v1/module/attention/dsa_mla.py (hunk 2 above)
Comment (sjyouuuuug) [reply xtuner/v1/module/attention/dsa_mla.py:564]: 现在只训练 source layer 的 indexer，后面只复用，不算 kl

File: xtuner/v1/module/attention/dsa_mla.py (hunk 1 above)
Comment (sjyouuuuug) [reply xtuner/v1/module/attention/dsa_mla.py]: 已经统一了，现在 weights_proj 只计算一次，并在 DSAIndexerFeatures 中保存一份 raw weights，不再维护 selection_weights 和 training_weights 两份张量。
